File: Sliding window/904. Fruit Into Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& tree) {
        unordered_map<int,int>mp;
        int cnt = 0, res = 0, start = 0;
        for(int i = 0; i<tree.size(); ++i){
            if(mp[tree[i]]++ == 0) ++cnt;
            while(cnt > 2){
                if(--mp[tree[start++]] == 0) --cnt;
            }
            if(cnt <= 2)
                res = max(res, i - start + 1);
        }
        return res;
    }
};

File: Sliding window/904. Fruit Into Baskets.py
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        res = cur = count_b = a = b = 0
        for c in tree:
            cur = cur + 1 if c in (a, b) else count_b + 1
            count_b = count_b + 1 if c == b else 1
            if b != c: a, b = b, c
            res = max(res, cur)
        return res
        
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        count, i = {}, 0
        for j, v in enumerate(tree):
            count[v] = count.get(v, 0) + 1
            if len(count) > 2:
                count[tree[i]] -= 1
                if count[tree[i]] == 0: del count[tree[i]]
                i += 1
        return j - i + 1
        
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        ans = i = 0
        count = collections.Counter()
        for j, x in enumerate(tree):
            count[x] += 1
            while len(count) >= 3:
                count[tree[i]] -= 1
                if count[tree[i]] == 0:
                    del count[tree[i]]
                i += 1
            ans = max(ans, j - i + 1)
        return ans
        